### ISSUE TITLE
Complex and real results do not agree when computing reciprocal or pow(-1) of 0

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/pytorch#45690]
## 🐛 Bug

When using a float or double input, the reciprocal or `pow(-1)` of 0 is infinity. However, when the input is complex, the result is NaN. Perhaps this is intended, but it seems wrong to me.

## To Reproduce

```
>>> torch.tensor(0., dtype=torch.float).pow(-1)
tensor(inf)
>>> torch.tensor(0., dtype=torch.cfloat).pow(-1)
tensor(nan+nanj)
>>> torch.tensor(0., dtype=torch.double).pow(-1)
tensor(inf, dtype=torch.float64)
>>> torch.tensor(0., dtype=torch.cdouble).pow(-1)
tensor(nan+nanj, dtype=torch.complex128)
```

## Expected behavior

I assume the reciprocal of 0 should always be the same, regardless of whether the type is real or complex.

**Comment by anjali411** [comment]
hmm I agree that one would probably expect to get `inf` whether the input is 0 or 0 + 0j. Currently, numpy and torch have the same behavior for reciprocaly of `0+0j`, i.e., both return `(nan+nanj)`. 
```
>>> import jax.numpy as np
>>> np.reciprocal(0.0)
DeviceArray(inf, dtype=float32)
>>> np.reciprocal(0.0+0.0j)
DeviceArray(inf+nanj, dtype=complex64)

>>> import numpy
>>> numpy.reciprocal(0.0)
__main__:1: RuntimeWarning: divide by zero encountered in reciprocal
inf
>>> numpy.reciprocal(0.0+0.0j)
__main__:1: RuntimeWarning: invalid value encountered in reciprocal
(nan+nanj)
>>>
```

cc. @mruberry

**Comment by kshitij12345** [comment]
Interestingly,
```python
>>> import numpy as np
>>> a = np.array(0+0j)
>>> a
array(0.+0.j)
>>> 1/a
(inf+nanj)
>>> np.reciprocal(a)
__main__:1: RuntimeWarning: invalid value encountered in reciprocal
(nan+nanj)
```

**Comment by kurtamohler** [comment]
Thanks for pointing that out @kshitij12345 . Perhaps this is a case where numpy shouldn't necessarily be followed? I submitted a numpy issue to see what they think: https://github.com/numpy/numpy/issues/17425

**Comment by mruberry** [comment]
Sidenote: in the near future we may also want to update torch.reciprocal to promote integer inputs to floats, so torch.reciprocal(5) = .2, unlike NumPy's reciprocal where np.reciprocal(5) = 0.

**Comment by anjali411** [comment]
> Aside: in the near future we may also want to update torch.reciprocal to promote integer inputs to floats, so torch.reciprocal(5) = .2, unlike NumPy's reciprocal where np.reciprocal(5) = 0.

so we would like to return a float tensor for `torch.reciprocal(int_tensor)`?

**Comment by mruberry** [comment]
> > Aside: in the near future we may also want to update torch.reciprocal to promote integer inputs to floats, so torch.reciprocal(5) = .2, unlike NumPy's reciprocal where np.reciprocal(5) = 0.
> 
> so we would like to return a float tensor for `torch.reciprocal(int_tensor)`?

Yes. That's the current thinking, anyway, but it hasn't been high priority.

**Comment by anjali411** [comment]
> Thanks for pointing that out @kshitij12345 . Perhaps this is a case where numpy shouldn't necessarily be followed? I submitted a numpy issue to see what they think: [numpy/numpy#17425](https://github.com/numpy/numpy/issues/17425)

From the discussion on the [numpy issue](https://github.com/numpy/numpy/issues/17425#issuecomment-702871715) it seems that `torch.tensor(0., dtype=torch.cfloat).reciprocal()` should return `tensor(inf+nanj)` since it `inf+nanj` is interpreted as `inf` which is the expected result for `1/0+0j` according to C99 spec. So we should fix our reciprocal behavior accordingly.

**Comment by kurtamohler** [comment]
@anjali411, it doesn't have to be exactly `inf+nanj`. The important thing is that the real part is infinity. The imaginary part can be anything. There might be a good argument for setting the imaginary part to a non-nan value. This way, we could convert the infinite result to a Python value and it would behave more nicely:

```
>>> 1.0 / (float('inf') + float('nan')*1j)
(nan+nanj)
>>> 1.0 / (float('inf') + 0j)
0j
```

Anyway, the numpy thread hasn't exactly reached a conclusion yet, so perhaps we should wait before taking any action.
